quinn-proto/src/connection.rs

```diff
@@ -875,37 +884,96 @@ impl Connection {
     }
 
     fn write_tls(&mut self) {
-        let mut outgoing = Vec::new();
-        self.tls.write_hs(&mut outgoing);
-        let offset = self.crypto_offset;
-        self.crypto_offset += outgoing.len() as u64;
-        if !outgoing.is_empty() {
-            self.crypto_pending.crypto.push_back(frame::Crypto {
-                offset,
-                data: outgoing.into(),
-            });
-            self.crypto_in_flight = true;
+        loop {
+            let space = self.highest_space;
+            let mut outgoing = Vec::new();
+            loop {
+                let prev_len = outgoing.len();
+                self.tls.write_hs(&mut outgoing);
+                match self.highest_space {
+                    SpaceId::Initial if self.spaces[SpaceId::Handshake as usize].is_none() => {
+                        if let Some(secrets) = self.tls.get_handshake_secrets() {
+                            self.upgrade_crypto(SpaceId::Handshake, secrets);
+                            // A client stops both sending and processing Initial packets when it
+                            // sends its first Handshake packet.
+                            if self.side.is_client() {
+                                self.spaces[SpaceId::Initial as usize] = None;
+                            }
+                            break;
+                        }
+                    }
+                    SpaceId::Handshake if self.spaces[SpaceId::OneRtt as usize].is_none() => {
+                        if let Some(secrets) = self.tls.get_1rtt_secrets() {
+                            self.upgrade_crypto(SpaceId::OneRtt, secrets);
+                            break;
+                        }
+                    }
+                    _ => {}
+                }
+                if outgoing.len() == prev_len {
+                    break;
+                }
+            }
+            if outgoing.is_empty() {
+                break;
+            }
+            let offset = self.crypto_offset;
+            self.crypto_offset += outgoing.len() as u64;
+            trace!(
+                self.log,
+                "wrote {} {space:?} TLS bytes",
+                outgoing.len(),
+                space = space
+            );
+            self.space_mut(space)
+                .unwrap()
+                .pending
+                .crypto
+                .push_back(frame::Crypto {
+                    offset,
+                    data: outgoing.into(),
+                });
         }
     }
 
+    /// Switch to stronger cryptography during handshake
+    fn upgrade_crypto(&mut self, space: SpaceId, secrets: rustls::quic::Secrets) {
```

Comment:
I would prefer to export `rustls` items through the `crypto` module, to maintain a bit more of an abstraction boundary -- the choice of `rustls` seems one of the more controversial aspects of our implementation, so it'd be nice to keep it wrapped up.

(I see `AlertDescription` already snuck in, but let's not make it worse?)